File: java/src/main/java/io/jkratz/katas/gameoflife/Board.java
package io.jkratz.katas.gameoflife;

/**
 * The {@code Board} class represents a two dimensional grid holding the
 * state of cells for Conway's Game of Life. The board is represent with
 * integers 0 representing a dead cell, and 1 representing a live cell.
 *
 * @author Joseph Kratz (joseph.kratz06@gmail.com)
 */
public class Board {

    public static int CELL_DEAD = 0;
    public static int CELL_ALIVE = 1;

    private int[][] state;
    private int rows;
    private int columns;

    /**
     * Creates a board instance with a random state for the size provided.
     *
     * @param rows The amount of rows on the board
     * @param columns The amount of columns on the board
     */
    public Board(int rows, int columns) {
        this.state = new int[rows][columns];
        for (int i=0; i<rows; i++) {
            for (int j=0; j<columns; j++) {
                this.state[i][j] = (int) Math.round(Math.random());
            }
        }
        this.rows = this.state.length;
        this.columns = this.state[0].length;
    }

    /**
     * Creates Board instance with a provided initial state. The initial state
     * must not be NULL, jagged, contain values other than 0 or 1, and must be
     * at least 1 row and 1 column.
     *
     * @param initialState The initial state of the board
     * @throws IllegalArgumentException If the {@code initialState} is not valid
     */
    public Board(int[][] initialState) {

        ValidationResult validationResult = validate(initialState);
        switch (validationResult) {
            case INVALID_BAD_VALUES:
            case INVALID_BAD_SIZE:
            case INVALID_NULL:
            case INVALID_JAGGED:
                throw new IllegalArgumentException(validationResult.message);
            default:
                this.state = initialState;
                this.rows = this.state.length;
                this.columns = this.state[0].length;
        }
    }

    /**
     * Returns the current state of the grid
     *
     * @return The current state
     */
    public int[][] getState() {
        return this.state;
    }

    /**
     * Returns the count of rows
     *
     * @return Rows on the board
     */
    public int rows() {
        return this.rows;
    }

    /**
     * Returns the count of columns
     *
     * @return Columns of on the board
     */
    public int columns() {
        return this.columns;
    }

    /**
     * Transitions to the next state of the Board.
     */
    public void evolve() {
        int[][] nextState = new int[this.rows][this.columns];

        // iterate through two dimensional array
        for (int i=0; i<rows; i++) {
            for(int j=0; j<columns; j++) {
                nextState[i][j] = this.getNextStateForCell(i,j);
            }
        }
        this.state = nextState;
    }

    /**
     * Gets next state for a particular cell
     *
     * @param i Row index of the grid
     * @param j Column index of the grid
     * @return Value for the next state (0 if dead, 1 if alive)
     */
    private int getNextStateForCell(int i, int j) {

        // Get living neighbors
        int aliveNeighbors = this.calculateLivingNeighbors(i,j);
        int cellValue = this.state[i][j];
        int newCellValue;

        if (cellValue == CELL_DEAD && aliveNeighbors == 3) {
            newCellValue = CELL_ALIVE;
        } else if (cellValue == CELL_ALIVE && (aliveNeighbors < 2 || aliveNeighbors > 3)) {
            newCellValue = CELL_DEAD;
        } else {
            newCellValue = cellValue;
        }

        return newCellValue;
    }

    /**
     * Returns the count of living neighbors for a particular cell in the grid
     *
     * @param i Row index of the grid
     * @param j Column index of the grid
     * @return Count of living neighbors for particular cell
     */
    private int calculateLivingNeighbors(int i, int j) {
        int liveCount = 0;
        for (int x=-1; x<=1; x++) {
            for (int y = -1; y <= 1; y++) {
                // check for boundary conditions
                if (i + x < 0 || i + x > (this.rows - 1) || y + j < 0 || y + j > (this.columns - 1)) {
                    continue;
                }
                liveCount += this.state[i + x][y + j];
            }
        }

        // remove since we may have counted ourselves
        liveCount -= this.state[i][j];
        return liveCount;
    }

    /**
     * Validates the state of the Board.
     *
     * @param state The state of the board
     * @return The result of the validation
     */
    private ValidationResult validate(int[][] state) {

        // If state is NULL return early as invalid, there is nothing else to check
        if (state == null) {
            return ValidationResult.INVALID_NULL;
        }

        if (state.length < 1 || state[0].length < 1) {
            return ValidationResult.INVALID_BAD_SIZE;
        }

        int size = state[0].length;
        for (int i=0; i<state.length; i++) {
            if (size != state[i].length) {
                return ValidationResult.INVALID_JAGGED;
            }
            for (int j=0; j<state[i].length; j++) {
                if (state[i][j] < 0 || state[i][j] > 1) {
                    return ValidationResult.INVALID_BAD_VALUES;
                }
            }
        }
        return ValidationResult.VALID;
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        for (int[] columns : this.state) {
            for (int value : columns) {
                if (value == 0) {
                    stringBuilder.append(" - ");
                } else {
                    stringBuilder.append(" * ");
                }
            }
            stringBuilder.append(System.lineSeparator());
        }
        return stringBuilder.toString();
    }

    /* Enum representing the possible validation results of the state of the board. */
    private enum ValidationResult {

        VALID(""),
        INVALID_JAGGED("Invalid input, state cannot be jagged"),
        INVALID_BAD_VALUES("Invalid values, only 0 and 1 are valid values"),
        INVALID_BAD_SIZE("There must be at least one row and one column"),
        INVALID_NULL("Input cannot be NULL");

        private String message;

        ValidationResult(String message) {
            this.message = message;
        }
    }
}

File: kotlin/src/main/kotlin/io/jkratz/katas/gameoflife/Board.kt
package io.jkratz.katas.gameoflife

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import java.util.*
import java.util.concurrent.CompletableFuture
import kotlin.collections.ArrayList

/**
 * Class to represent the state of the board for Conway's Game of Life.
 *
 * @author Joseph Kratz (joseph.kratz06@gmail.com)
 * @property state The state of the board
 * @property rows Count of rows on the board
 * @property columns Count of columns on the board
 */
class Board {

    var state: Array<IntArray>
        private set
    val rows: Int
    val columns: Int

    /**
     * Creates Board instance with a randomly generated state.
     */
    constructor(rows: Int, columns: Int) {
        this.state = Array(rows) { IntArray(columns) }
        for (i in 0 until rows) {
            for (j in 0 until columns) {
                this.state[i][j] = Math.round(Math.random()).toInt()
            }
        }
        this.columns = columns
        this.rows = rows
    }

    /**
     * Creates Board instance with an initial state.
     *
     * @param initialState The initial state, this cannot be a jagged array.
     * @throws IllegalArgumentException If the initialState is not valid
     */
    constructor(initialState: Array<IntArray>) {
        when (val validationResult = validate(initialState)) {
            ValidationResult.INVALID_BAD_SIZE,
            ValidationResult.INVALID_BAD_VALUES,
            ValidationResult.INVALID_JAGGED -> throw IllegalArgumentException(validationResult.message)
            else -> {
                this.state = initialState
                this.rows = this.state.size
                this.columns = this.state[0].size
            }
        }
    }

    /**
     * Transitions to the next state of the board.
     */
    fun evolve() {
//        runBlocking {
//            val nextState = Array(rows) { IntArray(columns) }
//            for (i in 0 until rows) {
//                for (j in 0 until columns) {
//                    nextState[i][j] = withContext(Dispatchers.Default) {
//                        getNextStateForCell(i, j)
//                    }
//                }
//            }
//            state = nextState
//        }
        val nextState = Array(rows) { IntArray(columns) }
        val futures = ArrayList<CompletableFuture<Unit>>()
        for (i in 0 until rows) {
            for (j in 0 until columns) {
                 futures.add(CompletableFuture.supplyAsync { getNextStateForCell(i,j) }
                    .thenApply {
                        nextState[i][j] = it
                    })
            }
        }
        futures.forEach { it.join() }
        state = nextState
    }

    /**
     * Gets next state for a particular cell
     *
     * @param i Row index of the grid
     * @param j Column index of the grid
     * @return 1 is cell alive, 0 if cell is dead
     */
    private fun getNextStateForCell(i: Int, j: Int): Int {

        val aliveNeighbors = calculateLivingNeighbors(i, j)
        val cellValue = state[i][j]

        return if (cellValue == CELL_ALIVE && (aliveNeighbors < 2 || aliveNeighbors > 3)) {
            CELL_DEAD
        } else if (cellValue == CELL_DEAD && aliveNeighbors == 3) {
            CELL_ALIVE
        } else {
            cellValue
        }
    }

    /**
     * Returns the count of living neighbors for a particular cell in the grid
     *
     * @param i Row index of the grid
     * @param j Column index of the grid
     * @return Count of living neighbors for particular cell
     */
    private fun calculateLivingNeighbors(i: Int, j: Int): Int {
        var liveCount = 0
        for (x in -1..1) {
            for (y in -1..1) {
                // check for boundary conditions
                if (i + x < 0 || i + x > rows - 1 || y + j < 0 || y + j > columns - 1) {
                    continue
                }
                liveCount += state[i + x][y + j]
            }
        }
        // remove since we may have counted ourselves
        liveCount -= state[i][j]
        return liveCount
    }

    /**
     * Validates the state of the board.
     *
     * @param state 2D state of the board
     */
    private fun validate(state: Array<IntArray>): ValidationResult {
        if (state.isEmpty() || state[0].isEmpty()) {
            return ValidationResult.INVALID_BAD_SIZE
        }

        val size = state[0].size
        for (i in 0 until state.size) {
            if (size != state[i].size) {
                return ValidationResult.INVALID_JAGGED
            }
            for (j in 0 until state[i].size) {
                if (state[i][j] < 0 || state[i][j] > 1) {
                    return ValidationResult.INVALID_BAD_VALUES
                }
            }
        }
        return ValidationResult.VALID
    }

    override fun toString(): String {
        return "Board(state=${Arrays.toString(state)}, rows=$rows, columns=$columns)"
    }

    companion object {

        const val CELL_ALIVE = 1
        const val CELL_DEAD = 0
    }

    /* Represents the validation result of the initial state */
    private enum class ValidationResult(val message: String) {
        VALID(""),
        INVALID_BAD_VALUES("The state contains invalid values, only 0 or 1 is allowed"),
        INVALID_BAD_SIZE("The state must be at least 1X1 (1 row and 1 column)"),
        INVALID_JAGGED("The state cannot be jagged.")
    }
}

/**
 * Prints the state of the board to standard out
 */
fun Board.prettyPrint() {
    for (columns in this.state) {
        for (value in columns) {
            if (value == 0) {
                print(" - ")
            } else {
                print(" * ")
            }
        }
        println()
    }
}
